planner/core/exhaust_physical_plans.go

```diff
@@ -769,26 +779,46 @@ func (p *LogicalJoin) constructInnerIndexScanTask(
 		cop.tablePlan = ts
 	}
 	is.initSchema(path.index, path.fullIdxCols, cop.tablePlan != nil)
-	rowSize := is.indexScanRowSize(path.index, ds, true)
-	sessVars := ds.ctx.GetSessionVars()
-	cop.cst = rowCount * rowSize * sessVars.ScanFactor
 	indexConds, tblConds := splitIndexFilterConditions(filterConds, path.fullIdxCols, path.fullIdxColLens, ds.tableInfo)
+	// Specially handle cases when input rowCount is 0, which can only happen in 2 scenarios:
+	// - estimated row count of outer plan is 0;
+	// - estimated row count of inner "DataSource + filters" is 0;
+	// if it is the first case, it does not matter what row count we set for inner task, since the cost of index join would
+	// always be 0 then;
+	// if it is the second case, HashJoin should always be cheaper than IndexJoin then, so we set row count of inner task
+	// to table size, to simply make it more expensive.
+	if rowCount <= 0 {
+		rowCount = ds.tableStats.RowCount
+	}
 	tmpPath := &accessPath{
 		indexFilters:     indexConds,
 		tableFilters:     tblConds,
+		countAfterIndex:  rowCount,
 		countAfterAccess: rowCount,
 	}
 	// Assume equal conditions used by index join and other conditions are independent.
+	if len(tblConds) > 0 {
+		selectivity, _, err := ds.tableStats.HistColl.Selectivity(ds.ctx, tblConds)
+		if err != nil || selectivity <= 0 {
+			logutil.BgLogger().Debug("unexpected selectivity, use selection factor")
+			selectivity = selectionFactor
+		}
+		tmpPath.countAfterIndex = rowCount / selectivity
+		tmpPath.countAfterAccess = tmpPath.countAfterIndex
+	}
 	if len(indexConds) > 0 {
 		selectivity, _, err := ds.tableStats.HistColl.Selectivity(ds.ctx, indexConds)
-		if err != nil {
-			logutil.BgLogger().Debug("calculate selectivity failed, use selection factor", zap.Error(err))
+		if err != nil || selectivity <= 0 {
+			logutil.BgLogger().Debug("unexpected selectivity, use selection factor")
 			selectivity = selectionFactor
 		}
-		tmpPath.countAfterIndex = rowCount * selectivity
+		tmpPath.countAfterAccess = tmpPath.countAfterIndex / selectivity
```

Comment:
The name `countAfterAccess` is calculated under `indexConds`, that looks a little strange.

Reply:
That is because we are backward reasoning from final row count.